subprojects/plugins/src/main/java/org/gradle/api/plugins/jvm/internal/DefaultJvmPluginServices.java

```diff
@@ -202,37 +190,41 @@ public ConfigurationVariant configureClassesDirectoryVariant(Configuration confi
-        return conf -> {
-            JavaPluginExtension javaPluginExtension = project.getExtensions().findByType(JavaPluginExtension.class);
-            if (alwaysEnabled || javaPluginExtension == null || !javaPluginExtension.getAutoTargetJvmDisabled()) {
-                int majorVersion = 0;
-                for (TaskProvider<COMPILE> compileTaskProvider : compileTasks) {
-                    COMPILE compileTask = compileTaskProvider.get();
-                    if (compileTask.getOptions().getRelease().isPresent()) {
-                        majorVersion = Math.max(majorVersion, compileTask.getOptions().getRelease().get());
-                    } else {
-                        int releaseFlag = getReleaseOption(compileTask.getOptions().getCompilerArgs());
-                        if (releaseFlag != 0) {
-                            majorVersion = Math.max(majorVersion, releaseFlag);
-                        } else {
-                            majorVersion = Math.max(majorVersion, Integer.parseInt(JavaVersion.toVersion(compileTask.getTargetCompatibility()).getMajorVersion()));
-                        }
-                    }
-                }
-                if (majorVersion != 0) {
-                    JavaEcosystemSupport.configureDefaultTargetPlatform(conf, majorVersion);
-                }
-            }
-        };
+    @Override
+    public <COMPILE extends AbstractCompile & HasCompileOptions> void useDefaultTargetPlatformInference(Configuration configuration, TaskProvider<COMPILE> compileTask) {
+        ConfigurationInternal configurationInternal = (ConfigurationInternal) configuration;
+        Set<TaskProvider<COMPILE>> compileTasks = Cast.uncheckedCast(configurationToCompileTasks.computeIfAbsent(configurationInternal, key -> new HashSet<>()));
+        compileTasks.add(compileTask);
+
+        JavaPluginExtension java = project.getExtensions().getByType(JavaPluginExtension.class);
+        configurationInternal.getAttributes().attributeProvider(
+            TargetJvmVersion.TARGET_JVM_VERSION_ATTRIBUTE,
+            providerFactory.provider(() -> getDefaultTargetPlatform(configuration, java, compileTasks))
+        );
     }
 
-    private static int getReleaseOption(List<String> compilerArgs) {
-        int flagIndex = compilerArgs.indexOf("--release");
-        if (flagIndex != -1 && flagIndex + 1 < compilerArgs.size()) {
-            return Integer.parseInt(String.valueOf(compilerArgs.get(flagIndex + 1)));
+    private static <COMPILE extends AbstractCompile & HasCompileOptions> int getDefaultTargetPlatform(Configuration configuration, JavaPluginExtension java, Set<TaskProvider<COMPILE>> compileTasks) {
+        assert !compileTasks.isEmpty();
+
+        if (!configuration.isCanBeConsumed() && java.getAutoTargetJvmDisabled()) {
+            return Integer.MAX_VALUE;
```

Comment:
How does this impact reports and errors? Do we end up seeing Integer.MAX_VALUE in errors?

Reply:
Yeah this is why I tried to avoid this solution initially. This will be visible anywhere we show attributes. So likely `println` from toString, any interaction in the debugger, variant matching exceptions printed to the console, and likely GE. 

The long term solution is to deprecate the `disableAutoTargetJvm` option entirely in favor of allowing the user to _override_ the value as opposed to set it to a blanket MAX_VALUE

Reply:
Another option would be to support empty (`null`) providers in the attribute container as meaning "no such attribute".
Since it would currently fail, that's a viable evolution.

Reply:
Been there, done that: https://github.com/gradle/gradle/pull/24738

The problem with this approach is that the attribute container inherently needs to call `isPresent` on the providers passed into it to properly support existing usages of the container. For example, users often write the following code:

```
for (Attribue<?> a : container.keySet()) {
    // `value` is assumed to be non-null!
    Object value = container.getAttribute(a)
}
```

Allowing non-present providers would break this assumption that _existing code already uses_, so we need to check if a given attribute is present before adding it to `keySet`.

Checking isPresent, particularly on task providers which have been `map`ped in some way, requires that provider to be realized. 

Because of this, we have determined that this is not a feasible approach, as it makes it way too easy to accidentally realize providers which provide values to the attribute container.